Build system improvements and fixes

guillemj opened the pull request:
Included build system and compiler warning fixes, and build system improvements, such as adding a version script to better control the exposed symbols from the shared library.

guillemj committed:
build: Remove obsolete AC_HEADERS_STDC macro

Recent autoconf versions have obsoleted this macro and emit a warning.

Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
test: Add missing braces around else branch

The code was properly indented but was missing the braces for the else
branch.

Warned-by: gcc
Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
build: Ignore backup files

Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
build: Add support for version script

This adds a new version script so that only public symbols are exported
and no private symbols leak. The first version node provided is based on
the first version that shipped with these symbols, and new version nodes
are expected to be added for each release adding them, linking the old
one to the new one.

Signed-off-by: Guillem Jover <guillem@hadrons.org>

kamalesh-babulal commented on lib/Makefile.am:
```diff
@@ -8,5 +8,9 @@ libsysfs_la_SOURCES = sysfs_utils.c sysfs_attr.c sysfs_class.c dlist.c \
+if HAVE_LINKER_VERSION_SCRIPT
+libsysfs_la_LDFLAGS += -Wl,--version-script=$(srcdir)/libsysfs.map
+endif
+EXTRA_libsysfs_la_DEPENDENCIES = libsysfs.map
```
Thanks for the improvements. I might be completely wrong here, should the `EXTRA_ibsysfs_la_DEPENDENCIES` moved into above if block? So that the dependency is only valid of the `HAVE_LINKER_VERSION_SCRIPT` is true

guillemj replied:
Ah nice catch, yes. While I don't think having the unconditional dependency hurts much as the file is included in the repo and  distribution, it is indeed strictly speaking not correct. :)

kamalesh-babulal commented on lib/libsysfs.map:
```diff
@@ -0,0 +1,104 @@
+	sysfs_read_dir_subdirs;
+	sysfs_remove_trailing_slash;
+	sysfs_write_attribute;
+} LIBSYSFS_PRIVATE;
```
Hope you have plans to add the symbols list for versions >= `2.1` too?

guillemj replied:
Ah sorry, it seems I messed up the conversion from the Debian symbols file to the version script, but now after verifying this again this seems to be a bit messy. AFAICS the content from git does not entirely match the released tarballs? I've gone by the latter, and added the new symbols in 2.1, will update also the Debian package as it seems I misplaced one there.

While going over this I've noticed few more things which I'm including in the rerolled patchset.

kamalesh-babulal replied:
Thanks for maintaining the package on Debian. I am assuming there is a mismatch with the `2.1` tag and tarball the distributions are shipping? Does the latest release `2.1.1` also has similar issue?.

guillemj replied:
Sorry was probably not very clear. One problem was that the Debian symbols file had a coupled of symbols introduced in 2.1 .0 marked as being introduced in 2.0.0, I've fixed those locally.

The problem with the mismatched upstream tarball and git is that the tags do not seem to be equivalent to the released tarballs on sourceforce. For example, for the sysfsutils-2_0_0 tag, there are none of the *_module_* symbols, but those are present in the released tarball. Or there is sysfs_open_device_tree() (supposedly introduced in 2.1.0) declared and used in the cmd/ dir, but there's no matching definition. Or there is sysfs_read_dir_subdirs (also supposedly introduced in 2.1.0) documented in the txt file but only present in the test suite.

I'm not sure whether these are artifacts of issues of the CVS to git conversion, or whether the CVS repo already contained mismatched content though?

Thanks for merging the changes!

kamalesh-babulal replied:
Thanks for fixing them. I am not sure what caused the mismatch, I will try and inspect them.

guillemj committed:
lib: Make add_subdirectory() static

This is a private function that should not be exported.

Warned-by: gcc
Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
build: Remove redundant EXTRA_CFLAGS definition

This variable will be set by autoconf on the Makefile.in as it is
declared as AC_SUBST in configure.ac.

Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
build: Fix typo in variable name

The variable was not being properly used as the name was typoed.

Signed-off-by: Guillem Jover <guillem@hadrons.org>

guillemj committed:
lib: Add a private prototype for sysfs_close_dev_tree()

This is a private wrapper function, which needs a prototype before its
definition, but only in the internal header file.

Warned-by: gcc
Signed-off-by: Guillem Jover <guillem@hadrons.org>
